### PR TITLE
Migrate docs to markdown and improve docstrings

### PR DESCRIPTION
**Pull request by ZeroCool940711**
This pull request includes the migration of documentation files from rst to markdown format and improves the docstrings for every function and class in the codebase. This will greatly enhance the documentation and overall code quality. This PR is meant to be improved before merging and is still a work in progress. Any mistakes such as typos or wrong examples should be fixed before we merge this PR.

### CODE REVIEW

**Comment by falkoschindler** [comment]
Hi @ZeroCool940711,

Thanks for this enormous pull request! It is certainly good to think about reStructuredText vs. Markdown and to extend the documentation here and there. But I also have some concerns:

1. Is Markdown really better? We all love Markdown and I personally would really like to switch. But is it as widely supported as reST? I heard that Sphinx expects reST? I just want to be sure before putting too much work into the conversion (probably too late anyway). We had a similar PR https://github.com/zauberzeug/nicegui/pull/1007 a few month ago and closed it due to missing support by PyCharm.

2. This PR is _huge_. I'd plead for splitting it up into (a) converting existing docstrings into reST, (b) adding more docstrings. Furthermore I'd like to concentrate on library code first (/nicegui), examples second and website as well as internal code last. Doing all at once makes it pretty hard for us to review everything in a reasonable perdiod of time.

3. This PR adds pretty verbose documentation. Although we try to document every public API, we're not convinced that _every_ function needs a docstring. We distinguish between documentation for the user (public API) and documentation for the developer. The latter has access to the source code and only needs _additional_ information. Ideally the code is self-explanatory.

    You also added class docstrings. We usually document classes in their initializers though. With both docstrings, I fear we have too much duplication, with one copy getting out of date when new methods are added or updated.

    And I think not every docstring needs fields like `Usage`, `Raises` or `Returns` if their content is trivial or `None`.

    Anyway, there's probably room for discussion. But we should do that _before_ changing the whole code base.

4. It looks like you're using a different (or no) formatter. In some classes you changed the indentation, creating a larger diff than necessary.

5. You seem to have replaced (some) single quote with double quotes. We try to stick with single quotes within the code base (apart from triple-double quotes for docstrings).

6. Last but not least, you introduced some issues by changing the code itself (e.g. `hashlib.md5` --> `md5`, duplicate `@dataclass` decorator) or removed comments marked with `NOTE`. I just quickly skimmed through a fraction of the diff. But such breaking changes let me loose trust in the PR. 😕

Long story short: First we need to find evidence that Markdown is widely supported in docstrings. Then we should convert existing docstrings from reST to Markdown, before discussing other fundamental changes to the documentation.

**Comment by ZeroCool940711** [comment]
> 1. Is Markdown really better? We all love Markdown and I personally would really like to switch. But is it as widely supported as reST? I heard that Sphinx expects reST? I just want to be sure before putting too much work into the conversion (probably too late anyway). 

I don't think markdown is 100% better than reST but I do think it will be easier to maintain and extend the documentation if we use Markdown for the documentation part. The issue is that the website is tied to the docstrings on the code, I first started to think of this PR because of the changes introduced in PR #2547 , as we can now see a snippet of the description when searching we need to improve the docstrings as much as possible to make the search and website more useful but there is the issue of NiceGUI not having a `ui.rest` while it has a `ui.markdown` which was used to show the description results when searching, so, I had no other choice but to think of moving the documentation to use Markdown instead of reST. I think some of the issues might be possible to solve if we introduce a `ui.rest` element or do something to change the way the descriptions are rendered, we might not even need to migrate to markdown if we have an element that can properly renter reST.

> 2. This PR is _huge_. I'd plead for splitting it up into (a) converting existing docstrings into reST, (b) adding more docstrings. Furthermore I'd like to concentrate on library code first (/nicegui), examples second and website as well as internal code last. Doing all at once makes it pretty hard for us to review everything in a reasonable perdiod of time.

I will look into splitting the PR into several PRs, I'm okay with doing everything from scratch if needed while doing it properly so you guys can review it, I understand that not everyone has as much free time as me.

> 3. This PR adds pretty verbose documentation. Although we try to document every public API, we're not convinced that _every_ function needs a docstring. We distinguish between documentation for the user (public API) and documentation for the developer. The latter has access to the source code and only needs _additional_ information. Ideally the code is self-explanatory.

I see, the main reason I tried to add as much information as possible was because of the user and the website, since the documentation is tied to the docstrings I thought that we should add more information to make it clear what things do and what they expect as well as what they return, this would make it easier to search for something by either using the website documentation or using an external search engine such as Google or Bing.

>    You also added class docstrings. We usually document classes in their initializers though. With both docstrings, I fear we have too much duplication, with one copy getting out of date when new methods are added or updated.
> And I think not every docstring needs fields like `Usage`, `Raises` or `Returns` if their content is trivial or `None`.

I first only added documentation to the `init` on each class and extended existing documentation with additional information, the reason for the `Raises` and `Returns` was with the idea of both the users and developers to have more information on what they can expect to get back when using something as well as potential issues they might find, if we have the `Raises` in the docstrings it will be easier to find and figure out why something is raising a certain error when using it incorrectly and it will also help with making it so it would be searchable on the website itself.

> 4. It looks like you're using a different (or no) formatter. In some classes you changed the indentation, creating a larger diff than necessary.
> 5. You seem to have replaced (some) single quote with double quotes. We try to stick with single quotes within the code base (apart from triple-double quotes for docstrings).

Sorry about that, seems like my global VSCode configuration took priority for some reason instead of the one on the workspace, I didn't notice until now since I'm used to using `Ruff` for everything which is my mistake, I will make sure to use the proper formatter from now on.

> 6. Last but not least, you introduced some issues by changing the code itself (e.g. `hashlib.md5` --> `md5`, duplicate `@dataclass` decorator) or removed comments marked with `NOTE`. I just quickly skimmed through a fraction of the diff. But such breaking changes let me loose trust in the PR. 😕

This issue is mainly because of Copilot which I used for helping with writing the docstrings, it seems to randomly change other stuff which is annoying, that specific error you mentioned as well with it changing the md5 code was something I noticed and had to fix, I'm pretty sure I commit the changes to fix that but I will look into that again, there was also the issue of it removing `@property` from some functions which broke a lot of stuff for me and I had to fix later, I mentioned this on the discord server to Rodja in a comment, it took me some time to figure out why nothing was sending events and buttons were not working.

> Long story short: First we need to find evidence that Markdown is widely supported in docstrings. Then we should convert existing docstrings from reST to Markdown, before discussing other fundamental changes to the documentation.

Don't worry, the whole point of this PR and why I made it as a draft was to discuss if it was good or not and fix any potential issues, I was aware that there would be issues with it and that it would need fixing.

**Comment by falkoschindler** [comment]
Let's close this PR for now and check if we find a solution for Markdown in docstrings that works in major IDEs. If so, we can create a new PR focusing on the conversion only before thinking about more verbose documentation.
